src/app/clusters/ota-server/ota-server.cpp

```diff
@@ -28,26 +28,81 @@
 
 #include <support/Span.h>
 
+#include "ota-server-delegate.h"
+#include "ota-server.h"
+
 using namespace chip;
+using namespace chip::app::clusters;
+
+using chip::app::clusters::OTAServerDelegate;
+
+namespace {
+constexpr uint8_t kLocationParamLength   = 2;   // The expected length of the Location parameter in QueryImage
+constexpr size_t kMaxMetadataLen         = 512; // The maximum length of Metadata in any OTA Server command
+constexpr size_t kUpdateTokenParamLength = 32;  // The expected length of the Update Token parameter used in multiple commands
 
-// TODO: all callbacks need implementations
+OTAServerDelegate * gDelegateTable[EMBER_AF_OTA_SERVER_CLUSTER_SERVER_ENDPOINT_COUNT] = { nullptr };
+
+OTAServerDelegate * GetDelegate(chip::EndpointId endpoint)
+{
+    uint8_t ep = emberAfFindClusterServerEndpointIndex(endpoint, ZCL_OTA_SERVER_CLUSTER_ID);
+    return (ep == 0xFF ? NULL : gDelegateTable[ep]);
+}
+
+bool SendStatusIfDelegateNull(chip::EndpointId endpointId)
+{
+    if (GetDelegate(endpointId) == nullptr)
+    {
+        ChipLogError(Zcl, "No OTAServerDelegate set for ep:%d", endpointId);
+        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_UNSUP_COMMAND);
+        return true;
+    }
+    return false;
+}
+} // namespace
 
 /**
  * @brief OTA Software Update Server Cluster ApplyUpdateRequest Command callback
  *
+ * @note It is the application's reponsibility to send the ApplyUpdateRequestResponse command after this is handled.
+ *
  * @param updateToken Identifier for the Software Image to be applied. Should be 32 octets long.
  * @param newVersion The SoftwareVersion value of the new Software Image that the client is ready to apply.
  */
 
 bool emberAfOtaSoftwareUpdateServerClusterApplyUpdateRequestCallback(app::Command * commandObj, chip::ByteSpan updateToken,
                                                                      uint32_t newVersion)
 {
-    return false;
+    EmberAfStatus status         = EMBER_ZCL_STATUS_SUCCESS;
+    chip::EndpointId endpointId  = emberAfCurrentEndpoint();
+    OTAServerDelegate * delegate = GetDelegate(endpointId);
+
+    ChipLogDetail(Zcl, "OTA Server received ApplyUpdateRequest");
+
+    if (SendStatusIfDelegateNull(endpointId))
+    {
+        return true;
+    }
+
+    if (updateToken.size() != kUpdateTokenParamLength)
+    {
+        // TODO: correct error?
+        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_INVALID_ARGUMENT);
```

Comment:
Suggest you make the server log errors with enough details to debug.

Reply:
Will do